### ISSUE TITLE
Vectorized operations over immutable structures (AKA Use of `@simd/@fast_math` with `NTuple`s)

### ISSUE THREAD

**Issue by jrevels** [JuliaDiff/NDuals.jl#1]
So I'm having a bit of trouble writing code that can take advantage of `@simd/@fast_math`, and it basically stems from the fact that `NTuple`s are immutable structures.

If you look at the most recent commit, I've replaced all the `map` and closure-reliant code with `@generated` functions. This improves performance quite a bit, especially for bivariate functions (speed up is roughly around the 200x range, and memory usage decreases by around 10x). Using the `@time` macro, it shows that only the minimum number of allocations necessary to perform the operation are used for most operations that I checked (e.g. `+{N}(::NDual{N}, ::NDual{N}` --> `N+1` allocations).

However, the strategy taken there was to simply inline the code into expressions of the form `tuple(arg1,arg2,...argN)` where the `arg`s are the appropriate results of the calculation. As you can see, this forgoes the use of `for` loops...can non-`for` loop code be auto-vectorized? My guess here is no, or at least not with Julia's auto-vectorizer, since I don't see how it could semantically distinguish linearly independent code snippets without a loop-like structure.

My intuition says to pre-allocate the container and then fill it with the appropriate values in a vectorized `for` loop, but obviously this can't be done using immutable containers like `NTuple`s. 

Any ideas here?

**Comment by scidom** [comment]
@jrevels can you elaborate a bit on what you mean by auto-vectorization of non-`for` loops?

**Comment by jrevels** [comment]
@scidom Basically I'm referring to Julia's ability to "automatically" vectorize code generated by the compiler, i.e. generate code that allows the processor to optimize linearly independent operations. For example, the `@simd` macro can be used on `for` loops in Julia to indicate that the loop body can be evaluated independently for each iteration (I found [this](https://software.intel.com/en-us/articles/vectorization-in-julia) to be a helpful write-up but I'm not sure how up-to-date it is). I'm actually not sure whether or not the `@simd` is an LLVM-based feature or not, though. 

I was doing a little bit of digging and came across [the auto-vectorization documentation for the LLVM](http://llvm.org/docs/Vectorizers.html) itself, which talks about an [SLP vectorizer](http://llvm.org/docs/Vectorizers.html#slp-vectorizer) that targets vectorization of non-loop code. It seems that [support for this vectorizer is available for use with Julia's tuples](https://github.com/JuliaLang/julia/pull/6271) but one has to use a flag when firing up the Julia REPL to trigger it. That issue was last updated in December, though, so I'm not sure if things changed in the recent tuple refactor.

I have *very* little experience with all of this, so my assumptions/knowledge might be wonky, but vectorization was mentioned in the project proposal @mlubin gave and I figured now would be a good to explore the details.

**Comment by jrevels** [comment]
[Julia's documentation on performance annotations](http://julia.readthedocs.org/en/latest/manual/performance-tips/#performance-annotations) is also a good resource for this topic.

**Comment by mlubin** [comment]
I think a reduced test case here would be a function that takes an ``NTuple{N,Float64}`` and a scalar and returns an ``NTuple{N,Float64}`` with the original components multiplied by the scalar. If we can get this working, then everything else is easy. May want to start a julia-users post on this.

**Comment by jrevels** [comment]
@mlubin Currently, I've defined such a function [here](https://github.com/JuliaDiff/NDuals.jl/blob/master/src/ndual.jl#L29), but I don't know if it can be written as a `for` loop such that we can leverage `@simd`.

A little off-topic, but It's worth noting that the current `NDual{4,T}` type implementation seems to perform exactly as well as `Dual4{T}` when comparing number of allocations, time, and memory usage for elementary operations (construction, `+`, `-`, `/`, `*`, etc.). The same appears to be true for `Dual{T}` and `NDual{1,T}` (and obviously, `NTuple{0,T}` is *more* performant than representing a purely real value in either of the other types). So it appears that the stack allocation of tuple members is providing the speed-up we expected. The only performance differences I see are for cases in which type conversion/promotion is involved, which `NDual` performs worse on.

**Comment by mlubin** [comment]
``@simd`` might not be exactly what we need. Also, a good way to compare ``Dual4{T}`` and ``NDual{4,T}`` is with ``code_llvm`` or ``code_native``. If they're the same, then you've done a good job :)

**Comment by jrevels** [comment]
@mlubin Ah, this is super useful. I caught some extra instructions in there for `/`, but I just pushed code that optimizes those away. Now the code generated for `exp`, `+`, `-`, `/`, `*` is exactly the same for `Dual{T}` vs. `NDual{1,T}` and `Dual4{T}` vs. `NDual{4,T}`. Initial construction is more costly by instruction count (e.g. `dual(...)` vs. `NDual(...)`), but performs about equivalently in time, allocation count, and memory usage when using `@time`.

**Comment by mlubin** [comment]
Nice! So now a question is if we can do the same (forgetting SIMD for a moment) without using generated functions. I don't know the answer here, may need to bug one of the core devs through julia-users.

**Comment by jrevels** [comment]
@mlubin What is the detriment to using generated functions to implement functions over tuple members (other than the SIMD issues)?

**Comment by mlubin** [comment]
It's not strictly necessary but would make the code much simpler. I dont think we can hope to replace DualNumbers completely with NDuals if the implementation is much less easy to understand. If you'd like to move on and look directly at the SIMD aspects, that's okay too.

**Comment by mlubin** [comment]
Related discussion: https://groups.google.com/forum/#!topic/julia-dev/uQ1DsGGSrD0

**Comment by davidanthoff** [comment]
I am sure that you've seen this, but julialang/julia#11902 would help a lot, right?

**Comment by jrevels** [comment]
@davidanthoff Thanks for pointing that out, I hadn't seen this yet! If the syntax discussed there is as performant as the equivalent inline code for constructing tuples, this would solve *so* many of our problems with the current `NDualTup` implementation (assuming that `@simd` can be used on such loops).

**Comment by davidanthoff** [comment]
Does [this](https://github.com/simonster/StructsOfArrays.jl) help? Probably not, seems a slightly different structure, right?

**Comment by jrevels** [comment]
@davidanthoff  I don't think has much application for what we talking about in this issue specifically, but it might be useful for when we evaluate large arrays of these objects over in ForwardDiff. I'm working on a [ridiculously over-large PR there](https://github.com/JuliaDiff/ForwardDiff.jl/pull/27) that, once merged, will basically make NDuals.jl and AutoDiffJSoC.jl obsolete.

**Comment by davidanthoff** [comment]
Great, hadn't seen that branch! This is ideal for me, right now I'm just using ForwardDiff.jl, so I'll just wait until you're done and will get some speedup.

Am I right that the SIMD/tuple thing is still not working, and won't without more support in Julia itself?

**Comment by jrevels** [comment]
Basically, yes. One of the key culprits is https://github.com/JuliaLang/julia/issues/11899, but even so, I believe it may still take some work before the SLP vectorizer is able to be triggered without starting up a special REPL session. The SLP vectorizer might not be necessary once https://github.com/JuliaLang/julia/pull/12113 is implemented, if support for `@simd` on loops over tuples becomes supported. However, I'm pretty sure that won't be merged to Base until after v0.4 is released. My lack of expertise in the area makes progress hard to follow, but I've heard that [the upcoming codegen rewrite](https://github.com/JuliaLang/julia/pull/11973) is also supposed to affect a lot of related things.

**Comment by davidanthoff** [comment]
Sounds good. I just read Arch's latest [comment](https://github.com/JuliaLang/julia/issues/11899) on JuliaLang/Julia#11899, that might require some tweaking down the road (thinks like tuple has to be of length 2, 4 or 8). But, all good stuff for Julia 0.5, I guess. I think what you achieved even without the SIMD stuff is great, I'm certainly looking forward to it.
